### PR TITLE
Added stat-mode

### LINKED ISSUE #6: Wrong stat
```js
var Decompress = require('decompress'),
	unzip = require('decompress-unzip');

new Decompress()
	.src('master.zip')
	.use(unzip({ strip: 1 }))
	.run(function (err, files) {
		files.forEach(function (file) {
			console.log(file.stat.isFile(), file.path);
		});
	});

```

[master.zip here](https://github.com/artem-malko/tars-handlebars/archive/master.zip)

### PR DESCRIPTION
**Pull request by TrySound**
fix #6

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -62,6 +63,7 @@ module.exports = function (opts) {
 				}
 
 				if (stat.isDirectory() || entry.fileName.charAt(entry.fileName.length - 1) === '/') {
+					new Mode(stat).isDirectory(true);
 					self.push(new File({
```
Comment (kevva) [inline index.js]:
```js
new Mode(stat).isDirectory(true);

self.push(new File({
```

File: index.js [hunk 2]
```diff
@@ -86,6 +88,9 @@ module.exports = function (opts) {
 							return;
 						}
 
+						if( ! stat.isSymbolicLink()) {
```
Comment (kevva) [inline index.js]: `if(!stat.isSymbolicLink()) {`

File: index.js [hunk 3]
```diff
@@ -86,6 +88,9 @@ module.exports = function (opts) {
...
+							new Mode(stat).isFile(true);
+						}
 						self.push(new File({
```
Comment (kevva) [inline index.js]:
```js
}

self.push(new File({
```

**Comment by kevva** [comment]
What does `stat.mode` produce if `mode` isn't defined?

**Comment by TrySound** [comment]
@kevva zero

**Comment by kevva** [comment]
Yeah, then I don't see the use case for this?

**Comment by TrySound** [comment]
@kevva After zero it adds type bytes by isDirectory(true) or isFile(true)

**Comment by shinnn** [comment]
@kevva, I found stat-mode useful but this PR still needs some update to meet our requirements. (I'll post detailed comments later... :zzz:)

**Comment by TrySound** [comment]
@shinnn Can you give detailed comments for PR?

File: index.js [hunk 4]
```diff
@@ -62,6 +63,8 @@ module.exports = function (opts) {
 				}
 
 				if (stat.isDirectory() || entry.fileName.charAt(entry.fileName.length - 1) === '/') {
```
Comment (shinnn) [inline index.js]: Can you remiove `stat.isDirectory() || `? https://github.com/kevva/decompress-unzip/issues/6#issuecomment-89003178

File: index.js [hunk 5]
```diff
@@ -62,6 +63,8 @@ module.exports = function (opts) {
...
+					new Mode(stat).isDirectory(true);
```
Comment (shinnn) [inline index.js]:
```js
if (!mode) {
	new Mode(stat).isDirectory(true);
}
```

File: index.js [hunk 6]
```diff
@@ -86,6 +89,10 @@ module.exports = function (opts) {
 							return;
 						}
 
+						if (!stat.isSymbolicLink()) {
```
Comment (shinnn) [inline index.js]: if (!mode) {
    new Mode(stat).isFile(true);
}

File: index.js (hunk 5 above)
Comment (TrySound) [inline index.js]: yauzl didn't set correct mode to files. So, I want to be sure this directory will be a directory by stat.

Comment (shinnn) [inline index.js]: Oh, really? Can you show us `console.log` result of [this code](https://github.com/kevva/decompress-unzip/issues/6#issue-65966351)?

Comment (TrySound) [inline index.js]: Okey, it's true. But why it cannot set correct mode to any file?

Comment (shinnn) [inline index.js]:
```js
var Decompress = require('decompress');
var unzip = require('decompress-unzip');

new Decompress()
    .src('master.zip')
    .use(unzip({ strip: 1 }))
    .run(function (err, files) {
        files.forEach(function (file) {
            console.log(file.stat.mode, file.path);
        });
    });
```

```
undefined '.gitignore'
undefined 'markup'
undefined 'README.md'
undefined 'markup/modules'
undefined 'markup/modules/_template'
33261 'markup/modules/_template/_template.html'
undefined 'markup/modules/footer'
undefined 'markup/modules/head'
33261 'markup/modules/footer/footer.html'
undefined 'markup/modules/head/data'
undefined 'markup/modules/head/data/data.js'
undefined 'markup/pages'
33261 'markup/modules/head/head.html'
33261 'markup/pages/_template.html'
undefined 'tars'
33261 'markup/pages/index.html'
undefined 'tars/tasks'
33261 'tars/tasks/compile-templates.js'
```

It sometimes fails (`README.md` etc.) but sometime succeeds (`markup/pages/_template.html` etc.).

Comment (TrySound) [inline index.js]: Done

**Comment by shinnn** [comment]
And, can you remove [L56-L58](https://github.com/kevva/decompress-unzip/blob/06a5cd6f21dd64d4fa75ada6d995ec7580e96c57/index.js#L56-L58)?

**Comment by TrySound** [comment]
@shinnn No, stat-mode use stat instance, not just mode value.

File: index.js [hunk 7]
```diff
@@ -61,7 +62,11 @@ module.exports = function (opts) {
 					stat.mtime = entry.getLastModDate();
 				}
 
-				if (stat.isDirectory() || entry.fileName.charAt(entry.fileName.length - 1) === '/') {
+				if (entry.fileName.charAt(entry.fileName.length - 1) === '/') {
+					if (!stat.mode) {
```
Comment (shinnn) [inline index.js]: if (!mode) {

File: index.js [hunk 8]
```diff
@@ -86,6 +91,10 @@ module.exports = function (opts) {
 							return;
 						}
 
+						if (!stat.mode) {
```
Comment (shinnn) [inline index.js]: `if (!mode) {`

**Comment by shinnn** [comment]
This code checks `mode` [here](https://github.com/kevva/decompress-unzip/blob/5afefa7b05af744c7778b0fe68c380c72d029396/index.js#L66) and [here](https://github.com/kevva/decompress-unzip/blob/5afefa7b05af744c7778b0fe68c380c72d029396/index.js#L94), so L56-L58 is not needed anymore.

**Comment by shinnn** [comment]
@TrySound Can you check if the current version meets your requirements or not?

**Comment by TrySound** [comment]
@shinnn It works fine, now. Thanks!

**Comment by shinnn** [comment]
Ohhhh, looks good to me.

```
(mode / path)

32768 '.gitignore'
16384 'markup'
32768 'README.md'
16384 'markup/modules'
16384 'markup/modules/_template'
33261 'markup/modules/_template/_template.html'
16384 'markup/modules/footer'
16384 'markup/modules/head'
33261 'markup/modules/footer/footer.html'
16384 'markup/modules/head/data'
32768 'markup/modules/head/data/data.js'
16384 'markup/pages'
33261 'markup/modules/head/head.html'
33261 'markup/pages/_template.html'
16384 'tars'
33261 'markup/pages/index.html'
16384 'tars/tasks'
33261 'tars/tasks/compile-templates.js'
```

**Comment by TrySound** [comment]
@shinnn Cool!

**Comment by shinnn** [comment]
@TrySound OK, can you squash those 5 commits into a single commit `Use stat-mode to set correct file mode`?

**Comment by TrySound** [comment]
@shinnn Sure. How to do it?

**Comment by shinnn** [comment]
1. `git rebase -i HEAD^^^^^`
2. https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#squash-your-changes-optional

    ````
    pick ...
    pick ...
    pick ...
    pick ...
    pick ...
    ````
    ↓
    (edit)
    ↓
    ````
    pick ...
    squash ...
    squash ...
    squash ...
    squash ...
    ````

    Then save and close the file.
3. `git push -f`

**Comment by TrySound** [comment]
@shinnn Sorry. How to remove merge?

**Comment by shinnn** [comment]
```
git merge --squash master
```

**Comment by TrySound** [comment]
@shinnn 
```
$ git push origin HEAD:master
Username for 'https://github.com': TrySound
Password for 'https://TrySound@github.com': 
To https://github.com/TrySound/decompress-unzip.git
 ! [rejected]        HEAD -> master (non-fast-forward)
error: failed to push some refs to 'https://github.com/TrySound/decompress-unzip.git'
hint: Updates were rejected because a pushed branch tip is behind its remote
hint: counterpart. Check out this branch and integrate the remote changes
hint: (e.g. 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

**Comment by shinnn** [comment]
Landed in 3602ba20455f128e38b304dd3d6e741528c8b1a8  :)

**Comment by TrySound** [comment]
@shinnn Okey :)

**Comment by TrySound** [comment]
Publish release?

**Comment by shinnn** [comment]
`3.2.2`

**Comment by kevva** [comment]
Nice work guys!

**Pull request closed without merging**